--- a/grub-core/loader/i386/linux.c
+++ b/grub-core/loader/i386/linux.c
@@ -687,8 +687,11 @@ grub_cmd_linux (grub_command_t cmd __att
 		  using_linuxefi = 1;
 		  return GRUB_ERR_NONE;
 		}
-	      grub_dprintf ("linux", "linuxefi failed (%d)\n", grub_errno);
-	      goto fail;
+	      if (grub_errno == GRUB_ERR_BAD_SIGNATURE || grub_errno == GRUB_ERR_ACCESS_DENIED) {
+		grub_error(GRUB_ERR_BAD_OS, "unsigned or wrong signed kernel image");
+	        grub_dprintf ("linux", "linuxefi failed (%d)\n", grub_errno);
+	        goto fail;
+	      }
 	    }
 	}
     }
