--- a/grub-core/normal/menu.c
+++ b/grub-core/normal/menu.c
@@ -790,6 +790,7 @@ run_menu (grub_menu_t menu, int nested,
               *auto_boot = 0;
 	      return current_entry;
 
+#ifdef IGEL_DEBUG
 	    case GRUB_TERM_ESC:
 	      if (nested)
 		{
@@ -811,6 +812,7 @@ run_menu (grub_menu_t menu, int nested,
 		    grub_menu_entry_run (e);
 		}
 	      goto refresh;
+#endif
 
 	    case GRUB_TERM_CTRL | 'l':
 	      menu_fini ();
