--- a/grub-core/commands/efi/tpm.c
+++ b/grub-core/commands/efi/tpm.c
@@ -32,7 +32,9 @@ typedef TCG_PCR_EVENT grub_tpm_event_t;
 
 static grub_guid_t tpm_guid = EFI_TPM_GUID;
 static grub_guid_t tpm2_guid = EFI_TPM2_GUID;
+#if 0
 static grub_guid_t cc_measurement_guid = GRUB_EFI_CC_MEASUREMENT_PROTOCOL_GUID;
+#endif
 
 static grub_efi_handle_t *grub_tpm_handle;
 static grub_uint8_t grub_tpm_version;
@@ -222,6 +224,7 @@ grub_tpm2_log_event (grub_efi_handle_t t
   return grub_efi_log_event_status (status);
 }
 
+#if 0
 static void
 grub_cc_log_event (unsigned char *buf, grub_size_t size, grub_uint8_t pcr,
 		   const char *description)
@@ -265,6 +268,7 @@ grub_cc_log_event (unsigned char *buf, g
   if (status != GRUB_EFI_SUCCESS)
     grub_efi_log_event_status (status);
 }
+#endif
 
 grub_err_t
 grub_tpm_measure (unsigned char *buf, grub_size_t size, grub_uint8_t pcr,
@@ -273,13 +277,13 @@ grub_tpm_measure (unsigned char *buf, gr
   grub_efi_handle_t tpm_handle;
   grub_efi_uint8_t protocol_version;
 
-  grub_cc_log_event(buf, size, pcr, description);
+  //grub_cc_log_event(buf, size, pcr, description);
 
   if (!grub_tpm_handle_find (&tpm_handle, &protocol_version))
     return 0;
 
-  grub_dprintf ("tpm", "log_event, pcr = %d, size = 0x%" PRIxGRUB_SIZE ", %s\n",
-                pcr, size, description);
+  //grub_dprintf ("tpm", "log_event, pcr = %d, size = 0x%" PRIxGRUB_SIZE ", %s\n",
+  //              pcr, size, description);
 
   if (protocol_version == 1)
     return grub_tpm1_log_event (tpm_handle, buf, size, pcr, description);
