Index: igelworkdir/grub-core/disk/i386/pc/biosdisk.c
===================================================================
--- igelworkdir.orig/grub-core/disk/i386/pc/biosdisk.c
+++ igelworkdir/grub-core/disk/i386/pc/biosdisk.c
@@ -28,6 +28,7 @@
 #include <grub/err.h>
 #include <grub/term.h>
 #include <grub/i18n.h>
+#include <grub/env.h>
 
 GRUB_MOD_LICENSE ("GPLv3+");
 
@@ -291,6 +292,7 @@ grub_biosdisk_iterate (grub_disk_dev_ite
 {
   int num_floppies;
   int drive;
+  const char *probe_cd_drives_env = NULL;
 
   /* For hard disks, attempt to read the MBR.  */
   switch (pull)
@@ -311,11 +313,34 @@ grub_biosdisk_iterate (grub_disk_dev_ite
       return 0;
 
     case GRUB_DISK_PULL_REMOVABLE:
+      probe_cd_drives_env = grub_env_get("PROBE_CD_DRIVES");
       if (cd_drive)
 	{
 	  if (grub_biosdisk_call_hook (hook, hook_data, cd_drive))
 	    return 1;
 	}
+      else if (probe_cd_drives_env)
+	{
+        /* probe cd drive */
+        for (drive = 0x9f; drive < 0xf0; drive++)
+	  {
+            struct grub_biosdisk_cdrp *cdrp
+             = (struct grub_biosdisk_cdrp *) GRUB_MEMORY_MACHINE_SCRATCH_ADDR;
+
+            cd_drive = drive;
+            grub_memset (cdrp, 0, sizeof (*cdrp));
+            cdrp->size = sizeof (*cdrp);
+            cdrp->media_type = 0xFF;
+            if ((! grub_biosdisk_get_cdinfo_int13_extensions (cd_drive, cdrp))
+                 && ((cdrp->media_type & GRUB_BIOSDISK_CDTYPE_MASK)
+                    == GRUB_BIOSDISK_CDTYPE_NO_EMUL))
+              {
+               if (grub_biosdisk_call_hook (hook, hook_data, cd_drive))
+                 return 1;
+              }
+            cd_drive = 0;
+	  }
+	}
 
       /* For floppy disks, we can get the number safely.  */
       num_floppies = grub_biosdisk_get_num_floppies ();
