--- a/grub-core/kern/verifiers.c
+++ b/grub-core/kern/verifiers.c
@@ -160,9 +160,12 @@ grub_verifiers_open (grub_file_t io, enu
       goto fail;
     }
 
+  if (!grub_strstr("tpm", ver->name) || grub_strstr(io->name, "vmlinuz")) {
+
   err = ver->write (context, verified->buf, ret->size);
   if (err)
     goto fail;
+  }
 
   err = ver->fini ? ver->fini (context) : GRUB_ERR_NONE;
   if (err)
@@ -174,6 +177,10 @@ grub_verifiers_open (grub_file_t io, enu
   FOR_LIST_ELEMENTS_NEXT(ver, grub_file_verifiers)
     {
       enum grub_verify_flags flags = 0;
+
+      if (grub_strstr("tpm", ver->name) && !grub_strstr(io->name, "vmlinuz")) {
+	continue;
+      }
       err = ver->init (io, type, &context, &flags);
       if (err)
 	goto fail_noclose;
