diff -Naurp a/linux-6.6.38.igel/drivers/misc/ipts/receiver.h b/linux-6.6.38.igel/drivers/misc/ipts/receiver.h
--- a/linux-6.6.38.igel/drivers/misc/ipts/receiver.h	1970-01-01 01:00:00.000000000 +0100
+++ b/linux-6.6.38.igel/drivers/misc/ipts/receiver.h	2024-08-02 15:41:31.551491490 +0200
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * Copyright (c) 2016 Intel Corporation
+ * Copyright (c) 2020 Dorian Stoll
+ *
+ * Linux driver for Intel Precise Touch & Stylus
+ */
+
+#ifndef _IPTS_RECEIVER_H_
+#define _IPTS_RECEIVER_H_
+
+#include <linux/mei_cl_bus.h>
+
+void ipts_receiver_callback(struct mei_cl_device *cldev);
+
+#endif /* _IPTS_RECEIVER_H_ */
