diff -Naurp a/linux-6.6.38.igel/drivers/platform/surface/surface3-wmi.c b/linux-6.6.38.igel/drivers/platform/surface/surface3-wmi.c
--- a/linux-6.6.38.igel/drivers/platform/surface/surface3-wmi.c	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/drivers/platform/surface/surface3-wmi.c	2024-08-02 15:41:31.559491681 +0200
@@ -37,6 +37,15 @@ static const struct dmi_system_id surfac
 			DMI_MATCH(DMI_PRODUCT_NAME, "Surface 3"),
 		},
 	},
+#ifdef CONFIG_IGEL_SURFACE_SURFACE3_OEMB_FIX
+	{
+		.matches = {
+			DMI_MATCH(DMI_BIOS_VENDOR, "American Megatrends Inc."),
+			DMI_MATCH(DMI_SYS_VENDOR, "OEMB"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "OEMB"),
+		},
+	},
+#endif /* CONFIG_IGEL_SURFACE_SURFACE3_OEMB_FIX */
 #endif
 	{ }
 };
