diff -Naurp a/linux-6.6.38.igel/drivers/firmware/sysfb.c b/linux-6.6.38.igel/drivers/firmware/sysfb.c
--- a/linux-6.6.38.igel/drivers/firmware/sysfb.c	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/drivers/firmware/sysfb.c	2024-08-02 15:41:31.535491108 +0200
@@ -33,6 +33,9 @@
 #include <linux/platform_device.h>
 #include <linux/screen_info.h>
 #include <linux/sysfb.h>
+#ifdef CONFIG_IGEL_VMWGFX_FIX
+#include <linux/pci.h>
+#endif
 
 static struct platform_device *pd;
 static DEFINE_MUTEX(disable_lock);
@@ -84,7 +87,17 @@ static __init int sysfb_init(void)
 	sysfb_apply_efi_quirks();
 
 	/* try to create a simple-framebuffer device */
+#ifdef CONFIG_IGEL_VMWGFX_FIX
+	if (pci_get_device(0x15ad, 0x0405, NULL) ||
+	    pci_get_device(0x15ad, 0x0406, NULL)) {
+		compatible = 0;
+	} else {
+		compatible = sysfb_parse_mode(si, &mode);
+	}
+#else
 	compatible = sysfb_parse_mode(si, &mode);
+#endif
+
 	if (compatible) {
 		pd = sysfb_create_simplefb(si, &mode);
 		if (!IS_ERR(pd))
