diff -Naurp a/linux-6.6.38.igel/drivers/misc/ipts/resources.h b/linux-6.6.38.igel/drivers/misc/ipts/resources.h
--- a/linux-6.6.38.igel/drivers/misc/ipts/resources.h	1970-01-01 01:00:00.000000000 +0100
+++ b/linux-6.6.38.igel/drivers/misc/ipts/resources.h	2024-08-02 15:41:31.551491490 +0200
@@ -0,0 +1,17 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * Copyright (c) 2016 Intel Corporation
+ * Copyright (c) 2020 Dorian Stoll
+ *
+ * Linux driver for Intel Precise Touch & Stylus
+ */
+
+#ifndef _IPTS_RESOURCES_H_
+#define _IPTS_RESOURCES_H_
+
+#include "context.h"
+
+int ipts_resources_alloc(struct ipts_context *ipts);
+void ipts_resources_free(struct ipts_context *ipts);
+
+#endif /* _IPTS_RESOURCES_H_ */
