diff -Naurp a/linux-6.6.38.igel/drivers/gpu/vga/vga_switcheroo.c b/linux-6.6.38.igel/drivers/gpu/vga/vga_switcheroo.c
--- a/linux-6.6.38.igel/drivers/gpu/vga/vga_switcheroo.c	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/drivers/gpu/vga/vga_switcheroo.c	2024-08-02 15:41:31.547491394 +0200
@@ -438,12 +438,16 @@ find_active_client(struct list_head *hea
 bool vga_switcheroo_client_probe_defer(struct pci_dev *pdev)
 {
 	if ((pdev->class >> 16) == PCI_BASE_CLASS_DISPLAY) {
+#ifdef CONFIG_IGEL_APPLE_GMUX
+		if (apple_gmux_present() && !vgasr_priv.handler_flags)
+#else
 		/*
 		 * apple-gmux is needed on pre-retina MacBook Pro
 		 * to probe the panel if pdev is the inactive GPU.
 		 */
 		if (apple_gmux_present() && pdev != vga_default_device() &&
 		    !vgasr_priv.handler_flags)
+#endif
 			return true;
 	}
 
