diff -Naurp a/linux-6.6.38.igel/drivers/media/i2c/dw9719.c b/linux-6.6.38.igel/drivers/media/i2c/dw9719.c
--- a/linux-6.6.38.igel/drivers/media/i2c/dw9719.c	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/drivers/media/i2c/dw9719.c	2024-08-02 15:41:31.551491490 +0200
@@ -6,6 +6,8 @@
  * https://github.com/ZenfoneArea/android_kernel_asus_zenfone5
  */
 
+#include <asm/unaligned.h>
+
 #include <linux/delay.h>
 #include <linux/i2c.h>
 #include <linux/pm_runtime.h>
