diff -Naurp a/linux-6.6.38.igel/security/commoncap.c b/linux-6.6.38.igel/security/commoncap.c
--- a/linux-6.6.38.igel/security/commoncap.c	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/security/commoncap.c	2024-08-02 15:41:31.571491967 +0200
@@ -1440,6 +1440,13 @@ int cap_mmap_file(struct file *file, uns
 
 #ifdef CONFIG_SECURITY
 
+#ifdef CONFIG_IGEL_APPARMOR_DBUS_SUPPORT
+static struct lsm_id capability_lsmid __lsm_ro_after_init = {
+	.lsm  = "capability",
+	.slot = LSMBLOB_NOT_NEEDED
+};
+#endif
+
 static struct security_hook_list capability_hooks[] __ro_after_init = {
 	LSM_HOOK_INIT(capable, cap_capable),
 	LSM_HOOK_INIT(settime, cap_settime),
@@ -1463,8 +1470,13 @@ static struct security_hook_list capabil
 
 static int __init capability_init(void)
 {
+#ifdef CONFIG_IGEL_APPARMOR_DBUS_SUPPORT
+	security_add_hooks(capability_hooks, ARRAY_SIZE(capability_hooks),
+			   &capability_lsmid);
+#else
 	security_add_hooks(capability_hooks, ARRAY_SIZE(capability_hooks),
 				"capability");
+#endif
 	return 0;
 }
 
