diff -Naurp a/linux-6.6.38.igel/security/apparmor/include/path.h b/linux-6.6.38.igel/security/apparmor/include/path.h
--- a/linux-6.6.38.igel/security/apparmor/include/path.h	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/security/apparmor/include/path.h	2024-08-02 15:41:31.571491967 +0200
@@ -13,6 +13,9 @@
 
 enum path_flags {
 	PATH_IS_DIR = 0x1,		/* path is a directory */
+#ifdef CONFIG_IGEL_APPARMOR_DBUS_SUPPORT
+	PATH_SOCK_COND = 0x2,
+#endif
 	PATH_CONNECT_PATH = 0x4,	/* connect disconnected paths to / */
 	PATH_CHROOT_REL = 0x8,		/* do path lookup relative to chroot */
 	PATH_CHROOT_NSCONNECT = 0x10,	/* connect paths that are at ns root */
