diff -Naurp a/linux-6.6.38.igel/security/apparmor/Makefile b/linux-6.6.38.igel/security/apparmor/Makefile
--- a/linux-6.6.38.igel/security/apparmor/Makefile	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/security/apparmor/Makefile	2024-08-02 15:41:31.567491872 +0200
@@ -7,6 +7,7 @@ apparmor-y := apparmorfs.o audit.o capab
               path.o domain.o policy.o policy_unpack.o procattr.o lsm.o \
               resource.o secid.o file.o policy_ns.o label.o mount.o net.o \
               policy_compat.o
+apparmor-$(CONFIG_IGEL_APPARMOR_DBUS_SUPPORT) += af_unix.o
 apparmor-$(CONFIG_SECURITY_APPARMOR_HASH) += crypto.o
 
 obj-$(CONFIG_SECURITY_APPARMOR_KUNIT_TEST) += apparmor_policy_unpack_test.o
