diff -Naurp a/linux-6.6.38.igel/security/apparmor/crypto.c b/linux-6.6.38.igel/security/apparmor/crypto.c
--- a/linux-6.6.38.igel/security/apparmor/crypto.c	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/security/apparmor/crypto.c	2024-08-02 15:41:31.571491967 +0200
@@ -25,6 +25,27 @@ unsigned int aa_hash_size(void)
 	return apparmor_hash_size;
 }
 
+#ifdef CONFIG_IGEL_APPARMOR_DBUS_SUPPORT
+void aa_snprint_hashstr(char *out, unsigned char *hash, unsigned int hsize)
+{
+       unsigned int i;
+
+       for (i = 0; i < hsize; i++)
+               sprintf(out + i*2, "%.2x", hash[i]);
+       out[hsize*2] = 0;
+}
+
+char *aa_asprint_hashstr(unsigned char *hash, unsigned int hsize, gfp_t gfp)
+{
+	char *buffer = kmalloc(hsize*2 + 1, gfp);
+	if (!buffer)
+		return NULL;
+	aa_snprint_hashstr(buffer, hash, hsize);
+
+	return buffer;
+}
+#endif
+
 char *aa_calc_hash(void *data, size_t len)
 {
 	SHASH_DESC_ON_STACK(desc, apparmor_tfm);
