diff -Naurp a/linux-6.6.38.igel/security/loadpin/loadpin.c b/linux-6.6.38.igel/security/loadpin/loadpin.c
--- a/linux-6.6.38.igel/security/loadpin/loadpin.c	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/security/loadpin/loadpin.c	2024-08-02 15:41:31.571491967 +0200
@@ -208,6 +208,13 @@ static int loadpin_load_data(enum kernel
 	return loadpin_check(NULL, (enum kernel_read_file_id) id);
 }
 
+#ifdef CONFIG_IGEL_APPARMOR_DBUS_SUPPORT
+static struct lsm_id loadpin_lsmid __lsm_ro_after_init = {
+	.lsm  = "loadpin",
+	.slot = LSMBLOB_NOT_NEEDED
+};
+#endif
+
 static struct security_hook_list loadpin_hooks[] __ro_after_init = {
 	LSM_HOOK_INIT(sb_free_security, loadpin_sb_free_security),
 	LSM_HOOK_INIT(kernel_read_file, loadpin_read_file),
@@ -259,7 +266,13 @@ static int __init loadpin_init(void)
 	if (!register_sysctl("kernel/loadpin", loadpin_sysctl_table))
 		pr_notice("sysctl registration failed!\n");
 #endif
+
+#ifdef CONFIG_IGEL_APPARMOR_DBUS_SUPPORT
+	security_add_hooks(loadpin_hooks, ARRAY_SIZE(loadpin_hooks),
+			   &loadpin_lsmid);
+#else
 	security_add_hooks(loadpin_hooks, ARRAY_SIZE(loadpin_hooks), "loadpin");
+#endif
 
 	return 0;
 }
