diff -Naurp a/linux-6.6.38.igel/sound/sound_core.c b/linux-6.6.38.igel/sound/sound_core.c
--- a/linux-6.6.38.igel/sound/sound_core.c	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/sound/sound_core.c	2024-08-02 15:41:31.579492158 +0200
@@ -251,7 +251,17 @@ retry:
 	
 	if (r < 0)
 		goto fail;
+#ifdef CONFIG_IGEL_CHANGE_OSS_DEVICE_NAMES
+	/* lang@igel: for the oss devices dsp/audio/mixer we want
+	   a number "0" at the end of the device name, because we create
+	   a device node link dsp -> dsp0 */
+	else if (r < SOUND_STEP &&
+		 strncmp(name, "dsp", 3) != 0 &&
+		 strncmp(name, "audio", 5) != 0 &&
+		 strncmp(name, "mixer", 5) != 0)
+#else
 	else if (r < SOUND_STEP)
+#endif
 		sprintf(s->name, "sound/%s", name);
 	else
 		sprintf(s->name, "sound/%s%d", name, r / SOUND_STEP);
