diff -Naurp a/linux-6.6.38.igel/tools/power/x86/intel-speed-select/isst-config.c b/linux-6.6.38.igel/tools/power/x86/intel-speed-select/isst-config.c
--- a/linux-6.6.38.igel/tools/power/x86/intel-speed-select/isst-config.c	2024-07-09 11:44:29.000000000 +0200
+++ b/linux-6.6.38.igel/tools/power/x86/intel-speed-select/isst-config.c	2024-08-02 15:41:31.579492158 +0200
@@ -4,11 +4,15 @@
  * Copyright (c) 2019 Intel Corporation.
  */
 
-#include <linux/isst_if.h>
-#include <sys/utsname.h>
-
+/* CONFIG_IGEL_WORKAROUND
+ * Makefile script fails and isst_if.h symbolic link is not created
+ * instead the reference to header file is in isst.h
+ * #include <linux/isst_if.h>
+ */
 #include "isst.h"
 
+#include <sys/utsname.h>
+
 struct process_cmd_struct {
 	char *feature;
 	char *command;
